### DISCUSSION TITLE
Setting Browser Cookie When Doing Server Side Rendering

### DISCUSSION THREAD

**Discussion by katiesun138** [vercel/next.js#53855] [Help]
###

Hi everyone - beginner Nextjser here!

I've been trying to create an application that has a login functionality.

Essentially if the user clicks on the form submit button (app/login/page.tsx), there will be a fetch request (POST) made to the API path (app/api/login/route.ts) that checks the database to verify that the user is in the database and to record the timestamp of when that user logged in. After verifying, I create a JWT token and I want to return that token to be stored in the browser cookie (so that in the future whenever I make a request from any page, the middleware will check if that token exists (to validate that the user is logged in).

Problem: I have created up to the JWT token step. But I'm having issues storing the token in the browser. When I use cookies().set(key, value) in the api path, the cookie is created, but not on the browser side. When I use cookies().set(key, value) in the page file, I get a "Cannot set headers after they are sent to the client". Any tips on how to set cookies (not using nextAuth)

### Additional information

Example code:
App/login/page.tsx
```js
const Login = (request: NextRequest, res: NextApiResponse) => {
    async function handleSubmit(data: any) {
    "use server";
    const loginInfo = {
        tempUsername: String(data.get("tempusername")),
        tempPassword: String(data.get("temppassword"))
    };
    const response = await fetch(`${server}/api/login`, 
        {method: "POST",
        headers: {"Content-Type": "application/json",},
        credentials: "include",
        body: JSON.stringify({ loginInfo }),});
    if (response.ok) {
        let obj: { [key: string]: string } = {};
        let cookie = response.headers.get("set-cookie");
        cookies().set("jwt", "hi"); //gives me error of "Cannot set headers after they are sent to the client
```

App/api/login/route.ts
```js
export async function POST(req: NextRequest) { 
    //code that will verify if user exists in database 
    const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN!;
    const token = await signJWT(
      {
        sub: exist[0]["submitted_at"],
      },
      { exp: `${JWT_EXPIRES_IN}m` }
    );

    const tokenMaxAge = parseInt(JWT_EXPIRES_IN) * 60; 
    const cStore = cookies(); 
    if (!cStore.get("token")) { 
        cStore.set("token", token); 
    }
    const responses = NextResponse.json({}, { status: 200 });
    return responses;
}
```

**Comment by devjiwonchoi** [comment]
Next.js `cookies()` is a Request Cookie which means it is not meant to be stored on the browser.
This is why the `cookies().set()` only works on Server Actions and Route Handlers.

> Good to know: .set() is only available in a [Server Action](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions) or [Route Handler](https://nextjs.org/docs/app/building-your-application/routing/route-handlers).

To set a cookie, return a new `Response`.

```ts
// route.ts
export async function POST(req: NextRequest) { 
    //code that will verify if user exists in database 
    const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN!;
    const token = await signJWT(
      {
        sub: exist[0]["submitted_at"],
      },
      { exp: `${JWT_EXPIRES_IN}m` }
    );

    const tokenMaxAge = parseInt(JWT_EXPIRES_IN) * 60; 
    const cStore = cookies(); 
    if (!cStore.get("token")) { 
        cStore.set("token", token); 
    }
    return new Response('hello', {
      status: 200,
      headers: { 'Set-Cookie': `jwt=${jwtToken}` },
    })
}
```

Note that you can't `NextResponse.next().cookies` since it is only allowed in middleware.

### Reference

- [cookies](https://nextjs.org/docs/app/api-reference/functions/cookies#cookiessetname-value-options)
- [Route Handlers Cookies](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#cookies)
- [NextResponse cookies](https://nextjs.org/docs/app/api-reference/functions/next-response#cookies)
- [`NextResponse.next()` used in a App Route Handler](https://nextjs.org/docs/messages/next-response-next-in-app-route-handler#why-this-error-occurred)

**Reply by katiesun138** [reply]
Thank you for the reply! 
I've tried returning a new response back to my page.tsx. I can see the cookie in the HeaderList when I do a console.log(response.headers), but what can I do so that the cookie is available for all the pages in my application to use when they are doing a post/get request. For each request, I want to check in the middleware if the cookie/jwt token exists; if it does exist then they can proceed to the page. 

As of right now,  in the middleware, when I do a 
`let cookie = request.cookies.getAll()`
This will only return the cookies that is from my browser, not the cookie jwt token that I just returned from my api route.

Thanks in advance!

**Reply by devjiwonchoi** [reply]
Since `middleware.js` is called before the `route.js` file, you won't be able to read the cookie from `middleware` that **will be defined** in `route`.

Why don't you create a component like `hasJWT` that validates the cookie then import it on your `route`?

```typescript
// hasJWT.ts
import { cookies } from 'next/headers'

export function hasJWT() {
  return !!cookies().get('jwt')
}
```

```typescript
// route.ts
import { NextRequest, NextResponse } from 'next/server'
import { cookies } from 'next/headers'
// ...other imports
import { hasJWT } from '@/utils'

// Move these to constants?
const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN || '1' // Set default number
const tokenMaxAge = parseInt(JWT_EXPIRES_IN) * 60
const exp = `${JWT_EXPIRES_IN}m`

export async function POST() {
  // ...
  if (!hasJWT()) {
    const sub = exist[0]?.['submitted_at']
    const token = await signJWT({ sub }, { exp }) // Does `signJWT` has any error handling?
    cookies().set('jwt', token)
  }
  return NextResponse.json({ message: 'success' }, { status: 200 })
}
```

Note:  It gives me an error when using `has` instead of `get`. I'll investigate it.

**Reply by katiesun138** [reply]
Just tried your suggested implementation - this seems to work for only one request in the API route. For example, if I submit the form, it will return false for the "hasJWT function" initially. Then, I would set the cookie (I can verify that the cookie is set correctly because when I console logged the hasJWT() again, the token value is returned). 
However, when I make another fetch POST request in the same page.tsx (after the first POST request) and called the hasJWT() function, it would return back to me false. So essentially setting the cookie did not persist.

 Here are snippets of my code to hopefully clarify my issue:
 
```
const Login = (request: NextRequest) => {
  async function handleSubmit(data: any) {
    "use server";
    const loginInfo = {
      tempUsername: String(data.get("tempusername")),
      tempPassword: String(data.get("temppassword"))
    };
    const response = await fetch("${server}/api/login", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      credentials: "include",
      body: JSON.stringify({ loginInfo })
    });
    if (response.ok) {
      let cookie = response.headers.get("set-cookie");
      console.log(cookie);  //in console, I will see my actual JWT token ( e.g. jwt="..." )

      const secondResponse = await fetch("${server}/api/setcookie", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Set-cookie": cookie!,
        },
        credentials: "include",
        body: JSON.stringify({ cookieInfo }),
      });

    } else {
      console.log("ERROR");
    }
  }

  return (
    <>HTML CODE</>)
```
Technically don't need to do another POST request to "set the cookie", but in general, when I make any other requests, my hasJWT token will always return false.

Thank you in advance! (I've been stuck on cookie creations for a while and I think it's really helpful getting your feedback and suggestions)

**Reply by devjiwonchoi** [reply]
It is because in Next.js `fetch`, it returns a matching cached data by default.

> If there is a match and it is fresh, it will be returned from the cache.

Try to add `{ cache: 'no-store' | 'no-cache' }` option to your fetch.
Note that `no-store` and `no-cache` are identical options for `fetch` in Next.js.

```ts
const response = await fetch("${server}/api/login", {
  cache: 'no-store',
  method: "POST",
  headers: {
    "Content-Type": "application/json",
  },
  credentials: "include",
  body: JSON.stringify({ loginInfo })
});
```

P.S. Is there a reason you don't move `api/setcookie` features inside `api/login`? Looks like a double-fetch is unnecessary.

### Reference

- [`fetch` options.cache](https://nextjs.org/docs/app/api-reference/functions/fetch#optionscache)

**Reply by devjiwonchoi** [reply]
Note: `has` in route is

Resolved with #54112
Released in [v13.4.17](https://github.com/vercel/next.js/releases/tag/v13.4.17)

**Reply by katiesun138** [reply]
Thank you for the update! I was able to resolve the issue after playing around with setting cookies within a component. It seems like setting a cookie within the specific API routes nor within the page component worked. So my resolution was to create a "setCookie.tsx" component that deals with all the form submissions. So in the Login page, when the user clicks the submit on the form, it will call a "login" function from the "setCookie.tsx" file. That function will set the jwt token, and do a page redirection to the next page in the website. 

I also have my middle ware checking to see if there exists "jwt" token in the request header. If there is, then it will authorize my user to access certain pages, otherwise it will redirect back to the login page. 

If you don't mind, I have another issue regarding server/client rendering and [next-intl](https://next-intl-docs.vercel.app/docs/getting-started/app-router-server-components) language switching. 
Issue: 
I have a navbar component that will include another component (languageSwitch.tsx) that has a href to allow the user to switch between two languages. If I submit a form that redirects from Login to the Welcome page, then if I click on the href link, it would switch languages, however it would redirect me back to the login page (but now the login page is in another language). 
I took a closer look and it seems like the issue is that with redirect('/welcome'), the languageSwitch component does not detect that I have switched paths (from /en/login to /en/welcome). So after, whenever I click on the href link in the languageSwitch component, it uses the old path (goes to /fr/login instead of /fr/welcome). 
I tested with using "use client". This seemed to solve my issue with language changes, however whenever I go backwards to previous pages, my nextjs13.4 applications seems to be stuck loading continuously.

Here is my code for reference:

NavBar.tsx
```
import Link from "next-intl/link";
import LanguageSwitch from "./LanguageSwitch";

const NavBar = () => {
  return (
    <nav>
      <div className={"container"}>
        <div id="left">
          //img on left
        </div>
        <div id="right">
          {/* @ts-expect-error Server Component */}
          <LanguageSwitch />
        </div>
      </div>
    </nav>
  );
};
export default NavBar;
```

LanguageSwitch.tsx (use client):
```
"use client";
import Link from "next-intl/link";
import { usePathname } from "next/navigation";

export default async function LanguageSwitch() {
  let pathname = usePathname();       //returns back e.g: "/en/login"
  let language: string = pathname!.includes("/en") ? "Français" : "English";
  let locale = pathname!.includes("/en") ? "fr" : "en";
  return (
    <header>
      <Link href={pathname!.substring(3)} locale={locale}>
        {language}
      </Link>
    </header>
  );
}
```

OR 

LanguageSwitch.tsx (use server):
```
import Link from "next-intl/link";
import { headers } from "next/headers";

export default async function LanguageSwitch() {
  const headerlist = await headers();                      
  const activepath = headerlist.get("x-invoke-path");          //returns back e.g: "/en/login"

  let language: string = activepath!.includes("/en") ? "Français" : "English";
  let locale = activepath!.includes("/en") ? "fr" : "en";

  return (
    <Link href={`${activepath!.substring(3)}`} locale={locale}>
      {language}
    </Link>
  );
}

```

Any suggestions would be helpful. If not, I can start up another discussion about this problem specifically. Thank you!
